File: LeetCode/NQueens2.cc
class Solution {
public:
    int totalNQueens(int n) {
        vector<int> perm(n);
        int ans;
        for(int i = 0; i < n; i++)
            perm[i] = i;
        do {
            if(check(perm)) {
                ans++;
            }
        } while(next_permutation(perm.begin(), perm.end()));
        return ans;
    }
    
    bool check(const vector<int> &perm) {
        for(int i = 0; i < perm.size(); i++)
            for(int j = 0; j < i; j++)
                if(abs(perm[i] - perm[j]) == i - j)
                    return false;
        return true;
    }
};

File: LeetCode/NQueens2.java
public class Solution {
    boolean used[];
    int ans;
    int[] perm;
    public int totalNQueens(int n) {
        used = new boolean[n];
        ans =  0;
        perm = new int[n];
        dfs(0);
        return ans;
    }
    public boolean checkDiagonal(int n) {
        for(int i = 0; i < n; i++)
            if(Math.abs(perm[i] - perm[n]) == n - i)
                return false;
        return true;
    }
    public void dfs(int n) {
        if(n == perm.length) {
            ans++;
            return;
        }
        for(int i = 0; i < perm.length; i++) {
            if(used[i]) continue;
            perm[n] = i;
            if(checkDiagonal(n)) {
                used[i] = true;
                dfs(n + 1);
                used[i] = false;
            }
        }
    }
}